### PR TITLE
Implemented unsized `Script`

### LINKED ISSUE #522: scripts should be a copy-on-write type rather than always owning their vec
Scripts are already read-only (we have the `Builder` type for modifying them) so it's quite limiting that there's no way to cheaply convert a byteslice as a script. For example you can't directly take a witness script off the witness stack to use it as a script_code in sighashcache.

### LINKED ISSUE #949: Reconsider the internal representation of `Script`
`Script` currently contains a boxed slice.

Pros/cons vs `Vec<u8>`:

* Faster to move (but moves are fast anyway, so the difference is crazy small)
* Consumes less memory: 1 `usize` less for capacity and guaranteed no extra memory
* Construction may force costly reallocation even for ridiculous values (e.g. to save 1 byte of memory)
* Having a shitload of scripts outside `Transaction` is unusual, the speed of moving `Transaction` is already independent of `Script` size (type; not to be confused with length)

Are the benefits worth extra reallocation on construction? My guess is no, but we should probably benchmark it.

Possible approaches:

* It can be demonstrated that the current design is really the best and we just close this issue
* It can be demonstrated `Vec<u8>` is the best and we switch
* Depends on application we expose feature flag to globally switch the strategy - easier but less flexible
* We make it generic over storage with some reasonable default

### PR DESCRIPTION
**Pull request by Kixunil**
This renames `Script` to `ScriptBuf` and adds unsized `Script` modeled
after `PathBuf`/`Path`. The change cleans up the API a bit, especially
all functions that previously accepted `&Script` now accept truly
borrowed version. Some functions that perviously accepted `&[u8]` can
now accept `&Script` because constructing it is no loger costly.

This is a more idiomatic alternative to #884 with two unavoidable lines of `unsafe` copied from `std`.

Closes #522 
Closes #949 

(For 949, we allow users to use whichever they like but still use `ScriptBuf` in `Transaction`.)

### CODE REVIEW

**Comment by tcharding** [comment]
There are a bunch of things I don't fully understand about this code but FWIW concept ACK from me, its a much cleaner that #884.

File: src/blockdata/script.rs
Comment (tcharding) [inline src/blockdata/script.rs:57]: As you work on this perhaps include this doc string modified from stdlib

```
+/// This is an unsized type, meaning that it must always be used behind a pointer like & or Box. For
+/// an owned version of this type, see [`ScriptBuf`].
```

Comment (Kixunil) [reply src/blockdata/script.rs:57]: Yeah, doc will be improved when I get more concept ACKs since there are two lines of `unsafe`.

**Comment by Kixunil** [comment]
Feel free to ask if you need help with understanding!

File: src/blockdata/script.rs
Comment (tcharding) [reply src/blockdata/script.rs:57]: no worries, just thought I'd throw it in since I was reading up trying to figure out what you were doing :)

**Comment by Kixunil** [comment]
Sadly, I'm [not 100% sure the casts are sound](https://internals.rust-lang.org/t/is-casting-dst-to-newtype-dst-actually-sound/17116). I believe the conclusion will be "it's OK" but until then it's formally UB.

File: src/blockdata/script.rs
Comment (elichai) [inline src/blockdata/script.rs:57]: Add `#[repr(transparent)]` to allow the casts?

Comment (Kixunil) [reply src/blockdata/script.rs:57]: I don't believe it helps. `#[repr(transparent)]` only affects the relationship between newtypes and their encapsulated types, not references to them. However this looks promising: https://github.com/rust-lang/reference/pull/1206

Comment (Kixunil) [reply src/blockdata/script.rs:57]: Thanks you! Looks like [`std` has the same issue](https://github.com/rust-lang/rust/blob/792bc5a0102d0973d42183a2b267850bb905236f/library/std/src/path.rs#L1162-L1167) however being `std` it can afford to do what it does and us being external library should do the opposite: use `#[repr(transparent)]` and document it's an implementation detail, right?

File: src/blockdata/script.rs
Comment (tcharding) [inline src/blockdata/script.rs:958]: My mind was just blown, this is acting like compile time polymorphism, right? This means that any method on `Script` can be called on a `ScriptBuf` because the compiler automatically dereferences the `ScriptBuf`, that is hot AF.

File: src/blockdata/script.rs
Comment (tcharding) [inline src/blockdata/script.rs:910]: Its not clear to me why these methods are on `ScriptBuf` when we have the `Builder`, can you explain the use case please?

**Review by tcharding — COMMENTED** [review]
There is a lot going on in this PR, I'm not sure my knowledge is deep enough to ack but I had a good long look and to the best of my understanding this looks pretty good.

File: src/blockdata/script.rs
Comment (Kixunil) [reply src/blockdata/script.rs:96]: Wanted to make `unsafe` separated and later I realized it's not needed here and forgot to merge it. :)

File: src/blockdata/script.rs
Comment (Kixunil) [reply src/blockdata/script.rs:654]: Ah, yes, I solved it in other case saying it's a method.

File: src/blockdata/script.rs
Comment (Kixunil) [reply src/blockdata/script.rs:821]: Remove them completely rather, I mindlessly moved things around.

File: src/blockdata/script.rs
Comment (Kixunil) [reply src/blockdata/script.rs:958]: Not really polymorphism, just `Deref` coercions. They work very well in case of `String` or `Vec` in my experience, so I went with it.

File: src/blockdata/script.rs
Comment (Kixunil) [reply src/blockdata/script.rs:910]: Yes, wanted to document better, but forgot. `Builder` has more ergonomic API for build-and-hand-over code but it requires `self`, where `ScriptBuf` takes `&mut self` instead.

File: src/blockdata/script.rs
Comment (apoelstra) [reply src/blockdata/script.rs:958]: I wouldn't use the word "just" to refer to `Deref` coercions :). I agree that this is one of the weirdest most magical parts of the language.

File: src/blockdata/script.rs
Comment (tcharding) [reply src/blockdata/script.rs:910]: Cool, I did not notice that detail. Thanks.

**Comment by Kixunil** [comment]
Review **not** addressed, just rebased for now. Couldn't find enough time. I'm half-sleeping rn.

**Comment by tcharding** [comment]
> I'm half-sleeping rn.

Hell of a team today then. I woke up in the middle of the night for some reason and started coding, pretty sure I will have done some brain fried stuff over the last few hours :)

File: src/lib.rs
Comment (Kixunil) [reply src/lib.rs:36]: Hmm, I just realized that by that logic we shouldn't have depended on `secp256k1` which has "buried `unsafe` code with implicit `allow`". I was thinking of moving `Script{,Buf,Builder}` to a separate crate to contain `unsafe` but this refutes the idea. (Could there be another good reason to move it though?)

**Comment by Kixunil** [comment]
Pushed more changes but review not addressed yet.

**Comment by Kixunil** [comment]
Last push addressed review comments and added a few more conversion APIs. Unfortunately, some required (also easy) `unsafe`. It's still not as good as `String`/`PathBuf`/... though, that will require a bunch more methods.

File: src/address.rs
Comment (tcharding) [inline src/address.rs:613]: Should be `Script` still.

File: src/address.rs
Comment (tcharding) [inline src/address.rs:66]: Should be `Script` still.

File: src/address.rs
Comment (tcharding) [inline src/address.rs:80]: Should be `Script` still.

File: src/blockdata/script.rs
Comment (tcharding) [inline src/blockdata/script.rs:124]: Do we need both these headings? If so, we could get rid of line 122

File: src/blockdata/script.rs
Comment (tcharding) [inline src/blockdata/script.rs:156]: Perhaps document the `unsafe`s.

File: src/blockdata/script.rs
Comment (tcharding) [inline src/blockdata/script.rs:901]: Perhaps `as_script_mut` is better? I think I read somewhere to put `mut` on the end but I failed to find a reference so not totally sure.

File: src/blockdata/script.rs
Comment (tcharding) [inline src/blockdata/script.rs:914]: With two types so similar in the same file perhaps its clearer to return `Script` instead of `Self`?

File: src/blockdata/script.rs
Comment (tcharding) [reply src/blockdata/script.rs:901]: Oh and we have `as_bytes` and `as_bytes_mut` so would be uniform with those.

File: src/util/psbt/map/input.rs
Comment (tcharding) [inline src/util/psbt/map/input.rs:36]: This and below should all be `Script`

File: src/address.rs
Comment (Kixunil) [reply src/address.rs:613]: LOL, I did search&replace and I didn't realize the compiler won't complain if the doc is wrong.

File: src/blockdata/script.rs
Comment (Kixunil) [reply src/blockdata/script.rs:156]: There's a clippy lint for that, we should activate it.

File: src/blockdata/script.rs
Comment (Kixunil) [reply src/blockdata/script.rs:901]: There's [as_mut_slice](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.as_mut_slice) I modeled it after that - it's the analogous function. I need to change `as_bytes_mut` instead.

File: src/blockdata/script.rs
Comment (Kixunil) [reply src/blockdata/script.rs:914]: I always use `Self` for return types to make refactorings less painful. Although ideally there won't be any such refactorings...

File: src/blockdata/script.rs
Comment (tcharding) [reply src/blockdata/script.rs:156]: Seems it only lints for unsafe functions, not functions that include unsafe blocks internally. Its on by default.

File: src/blockdata/script.rs
Comment (tcharding) [reply src/blockdata/script.rs:901]: Cool, TIL. `as_mut_ptr` exists too. Cheers.

File: src/blockdata/script.rs
Comment (stevenroose) [inline src/blockdata/script.rs:155]: I never noticed how this was possible! So there is a way to do this with unsafe! Amazing!

File: src/consensus/encode.rs
Comment (stevenroose) [inline src/consensus/encode.rs:676]: I would avoid doing this. It makes it quite error-prone, because it's unclear if slices would be length-prefixed or not. If we use this internally, I'd prefer using `consensus_encode_with_size` instead..

Comment (stevenroose) [reply src/consensus/encode.rs:676]: Generally, we just use the `emit_slice` method from `WriteExt`.

Comment (Kixunil) [reply src/consensus/encode.rs:676]: Good point, though IIRC `Vec<u8>` implements `Encodable` and it shouldn't for the same reason.

**Comment by tcharding** [comment]
Whats the status of this one @Kixunil?

**Comment by Kixunil** [comment]
Definitely want to continue but since we were doing other things I didn't have much time. Now this will need an insane rebase, unfortunately.

**Review by sanket1729 — COMMENTED** [review]
Did a pass-through. Looks good to me, light code review ACK.

File: src/blockdata/script.rs
Comment (Kixunil) [reply src/blockdata/script.rs:1247]: Wow, I completely forgot I wrote this. :D I believe I've looked at common scripts and picked the largest which is indeed P2PKH.

**Comment by Kixunil** [comment]
I've added milestone as I'm hoping to get it into the next version. I would like to rebase this no later than during the next week.

**Comment by tcharding** [comment]
I see no reason we cannot get this in 0.30.0

**Comment by apoelstra** [comment]
I agree, I'd really like to get this into the next version.

File: bitcoin/src/blockdata/script.rs
Comment (tcharding) [inline bitcoin/src/blockdata/script.rs:1742]: TIL: Our pre-commit hook and clippy CI job allows one character of trailing whitespace - does that mean one character of trailing whitespace is ok, clippy is wrong, or we have the clippy incantation wrong?

`cargo clippy --all-features --all-targets -- -D warnings`

File: bitcoin/src/blockdata/script.rs
Comment (tcharding) [inline bitcoin/src/blockdata/script.rs:589]: What is the point of this impl? Is it just for uniformity (along with `as_mut`) so `ScriptBuf` and `Script` both have these implemented?

File: bitcoin/src/blockdata/script.rs
Comment (tcharding) [inline bitcoin/src/blockdata/script.rs:1251]: Can you explain to me how this works please? I was expecting to see `self.as_script()`. Same question for all the others (especially `AsRef`).

**Review by tcharding — APPROVED** [review]
Excluding a few minor things that I don't quite get, ACK a620afb0d767b380b505ccc8a5a51ba0b2212b3a

File: bitcoin/src/blockdata/script.rs
Comment (Kixunil) [reply bitcoin/src/blockdata/script.rs]: Oh weird, I though I removed all of them.

File: bitcoin/src/blockdata/script.rs
Comment (Kixunil) [reply bitcoin/src/blockdata/script.rs:615]: It allows you to pass `&Script` to functions requiring `AsRef<Script>`. `std` cannot implement this as a blanket impl due to coherence issues (it does impl `AsRef<T> for &T` though) so types need to do it themselves. This is opposite of `Borrow` which has blanket `Borrow<T> for T` but not for references.

[`std` types have this impl too](https://doc.rust-lang.org/stable/std/path/struct.Path.html#impl-AsRef%3CPath%3E-for-Path) so it's also consistent with that.

File: bitcoin/src/blockdata/script.rs
Comment (Kixunil) [reply bitcoin/src/blockdata/script.rs:1320]: Deref coercion :)

**Comment by Kixunil** [comment]
I went over the PR and noticed/fixed some more issues. One still remains: `Display` calls into `Debug` which outputs `Script(ASM)`. Firstly, calling `Debug` from `Display` is wrong. But what should `Script` even display? No representation (hex, asm) seems obvious enough.

Because of the confusion I'd prefer removing `Display` and adding `display_hex()` and `display_asm()` methods similar to how we handled amount.

**Comment by apoelstra** [comment]
I think `Display` should show ASM and we should use the `LowerHex`/`UpperHex` traits for hex. The reason being that while either hex or asm could be the "natural" choice for `Dispaly`, hex gets its own format specifier so we can just move it there.

I don't think it'll cause confusion in practice; output of `Display` is the *one* thing that programmers visually check :P.

**Comment by Kixunil** [comment]
Oh, yes, I'm dumb last couple of days (I'm a bit sick).
Should we put `Script()` around it though? Seems wrong for `Display`.

**Comment by apoelstra** [comment]
Agreed it seems wrong -- so I vote no on putting `Script` around it.

**Comment by apoelstra** [comment]
> I think `Display` should show ASM and we should use the `LowerHex`/`UpperHex` traits for hex. The reason being that while either hex or asm could be the "natural" choice for `Dispaly`, hex gets its own format specifier so we can just move it there.
> 
> I don't think it'll cause confusion in practice; output of `Display` is the _one_ thing that programmers visually check :P.

@Kixunil I think that if you implement this, and rebase, I can do a final review of this PR and we can finally get it in.

**Comment by Kixunil** [comment]
I noticed hex encoding is using the slow implementation. Since I'm low on time and this PR may block others I suggest to let it in and I will change to fast impl in a subsequent PR. Also I kept `Script()` in `Debug` impl. Again, if we want to change it I suggest a separate PR.

File: bitcoin/src/blockdata/script.rs
Comment (Kixunil) [inline bitcoin/src/blockdata/script.rs:612]: FYI this doc attribute doesn't work. I prefer to fix it in a separate PR.

**Comment by apoelstra** [comment]
Needs rebase again, sorry, I think because of #1330.

File: bitcoin/src/blockdata/script.rs
Comment (apoelstra) [inline bitcoin/src/blockdata/script.rs:1270]: nit: (can be dealt with in a followup PR). I don't understand this sentence.

File: bitcoin/src/blockdata/script.rs
Comment (apoelstra) [inline bitcoin/src/blockdata/script.rs:1282]: nit: we should mention here that we don't check whether `last_opcode` is actually the last opcode

File: bitcoin/src/blockdata/script.rs
Comment (Kixunil) [reply bitcoin/src/blockdata/script.rs:1270]: It's he issue I brought up before :) I will try to think of better wording after I sleep.

File: bitcoin/src/blockdata/script.rs
Comment (Kixunil) [reply bitcoin/src/blockdata/script.rs:1282]: This is a private method so it shouldn't matter that much but yeah, something for another PR.

Comment (apoelstra) [reply bitcoin/src/blockdata/script.rs:1282]: Oh, perfect :) for a private method it's fine, it's obvious from the code that it doesn't check.

File: bitcoin/src/blockdata/script.rs
Comment (apoelstra) [reply bitcoin/src/blockdata/script.rs:1270]: Maybe "Note that `VERIFY` itself is not a `*VERIFY` opcode -- repeated instances of `OP_VERIFY` will each pop a distinct stack element, so they cannot be combined."

Anyway we can debate this on a followup PR, let's not hold up a 2000-line diff on this :)

**Comment by apoelstra** [comment]
Wow, what a magnum opus! This is very thorough. I appreciate all the long-form documentation as well as the many utility methods that make `ScriptBuf` ergonomic. ACK.

**Review by apoelstra — APPROVED** [review]
ACK 8e428562cbff89aa3f9d111792eb6105cfd8cf03

**Review by tcharding — APPROVED** [review]
ACK 8e428562cbff89aa3f9d111792eb6105cfd8cf03

File: bitcoin/src/lib.rs
Comment (tcharding) [inline bitcoin/src/lib.rs:62]: I don't fully understand the cost/benefits of always using the `alloc` crate vs using std for std builds and alloc only for no-std builds. Not important for this PR but its something I've noticed I didn't understand this week, if you have time ever can you explain your thoughts please?

Comment (tcharding) [reply bitcoin/src/lib.rs:62]: Oh, needs more context, the lines below this are 
```rust
#[macro_use]
extern crate alloc;
```

**Comment by sanket1729** [comment]
Please wait before merging this. On top of my list for reviews today. Should be done in a couple of hours.

**Review by sanket1729 — APPROVED** [review]
ACK 8e428562cbff89aa3f9d111792eb6105cfd8cf03. Let's get this in

**Comment by sanket1729** [comment]
I think 3 ACKs should this good to go. Merging this to prevent other accidental merges from breaking this big PR

**Comment by sanket1729** [comment]
and BOOM!

File: bitcoin/src/lib.rs
Comment (Kixunil) [reply bitcoin/src/lib.rs:62]: > I don't fully understand the cost/benefits of always using the alloc crate vs using std for std builds and alloc only for no-std builds. 

Just simpler code, less compilation conditions to worry about.

**Comment by Kixunil** [comment]
Thanks a lot folks for quick review, I didn't expect it!

**Pull request merged**
